internal/defence/limiter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package defence
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"storj.io/storj/internal/sync2"
+)
+
+// Attacker stores information about attacker entity
+type Attacker struct {
+	limiter *rate.Limiter
+	Expire  time.Time
+	// TODO: could be interface{}
+	key string
+
+	IsBanned bool
+}
+
+func (attacker *Attacker) attack(attemptTime time.Time, banDuration time.Duration) bool {
+	// is not cleared by Clear method of Limiter
+	if attacker.IsBanned && attemptTime.After(attacker.Expire) {
+		attacker.IsBanned = false
+	}
+
+	if attacker.IsBanned {
+		return true
+	}
+
+	isBanned := !attacker.limiter.AllowN(attemptTime, 1)
+
+	if isBanned {
+		attacker.IsBanned = true
+		attacker.Expire = attemptTime.Add(banDuration)
+	}
+
+	return isBanned
+}
+
+// Limiter is used to store and manage list of banned entities
+type Limiter struct {
+	attackers map[string]*Attacker
+
+	// Attempts defines how many times attacker could perform an operation
+	Attempts int
+	// AttemptsPeriod defines period in which attempts will count. For example, 5 attempts per minute.
+	AttemptsPeriod time.Duration
+	BanDuration    time.Duration
```

Comment:
any reason these are public?

Reply:
i tried to make this struct private, but linter return warnings:
`exported method Banned returns unexported type []*defence.attacker, which can be annoying to use`

Reply:
Yeah, the issue is that it probably shouldn't be returned.

Reply:
yeah, this fields should be private

Reply:
`BanDuration` doesn't look like a good name for what's used.